### PR TITLE
[TCGC] Fix judge logic how to distinguish model is referred both in common content-type and multipart content-type

### PR DESCRIPTION
**Pull request by msyyc**
fix https://github.com/Azure/typespec-azure/issues/1510

### CODE REVIEW

File: packages/typespec-client-generator-core/src/types.ts [hunk 1]
```diff
@@ -1677,6 +1657,27 @@ function updateTypesFromOperation(
...
     diagnostics.pipe(updateUsageOrAccessOfModel(context, access, sdkType));
+
+    // after complete of usage calculation for httpBody, check whether it has
+    // conflicting usage between multipart and regular body
+    if (
+      sdkType.kind === "model" &&
+      ((!multipartOperation && (sdkType.usage & UsageFlags.MultipartFormData) > 0) ||
+        (multipartOperation &&
+          (sdkType.usage & UsageFlags.MultipartFormData) > 0 &&
+          ((sdkType.usage & UsageFlags.Json) | (sdkType.usage & UsageFlags.Xml)) > 0))
```
Comment (msyyc) [inline packages/typespec-client-generator-core/src/types.ts:1668]: As discussed in https://github.com/Azure/typespec-azure/issues/1510#issuecomment-2345259187, we shall usage existing usage `xml/Json` to distinguish whether a model has conflicting usage between multipart operation and regular operation.
And we shall move the logic position after complete of usage calculation.

File: packages/typespec-client-generator-core/test/types/multipart-types.test.ts [hunk 2]
```diff
@@ -61,6 +61,22 @@ describe("typespec-client-generator-core: multipart types", () => {
...
   });
+  it("multipart conflicting model usage for only multipart operations", async function () {
+    await runner.compile(
+      `
+        @service({title: "Test Service"}) namespace TestService;
+
+        model MultiPartRequest {
+          id: string;
+          profileImage: bytes;
+        }
```
Comment (weidongxu-microsoft) [inline packages/typespec-client-generator-core/test/types/multipart-types.test.ts:72]: Should we add an `Address` model and verify usage of it is JSON and not MultipartFormData?

Also suggest we can verify the `MultiPartRequest` model be MultipartFormData and not JSON.

BTW, why to have a PUT and a POST in op?

Comment (msyyc) [reply packages/typespec-client-generator-core/test/types/multipart-types.test.ts]: _Should we add an Address model and verify usage of it is JSON and not MultipartFormData?
Also suggest we can verify the MultiPartRequest model be MultipartFormData and not JSON._

-> Updated test. And I also add another test for mixed operation scenarios

Comment (msyyc) [reply packages/typespec-client-generator-core/test/types/multipart-types.test.ts]: _BTW, why to have a PUT and a POST in op?_
-> With different method, there is no need to add route to distinguish them. Maybe it caused confusion, I have updated the test.

File: packages/typespec-client-generator-core/test/types/multipart-types.test.ts [hunk 3]
```diff
@@ -61,6 +61,68 @@ describe("typespec-client-generator-core: multipart types", () => {
...
+    deepEqual(multiPartRequest.usage, UsageFlags.MultipartFormData | UsageFlags.Input);
+  });
+  it("multipart conflicting model usage for mixed operations", async function () {
+    await runner.compile(
+      `
+        @service({title: "Test Service"}) namespace TestService;
+        model Address {city: string;}
+        model RegularRequest {
+          address: Address;
+        }
```
Comment (weidongxu-microsoft) [inline packages/typespec-client-generator-core/test/types/multipart-types.test.ts:100]: nit, does Address still have usage=JSON, if there is no RegularReqeust?

**Review by weidongxu-microsoft — APPROVED** [review]

File: packages/typespec-client-generator-core/test/types/multipart-types.test.ts [hunk 4]
```diff
@@ -61,6 +61,68 @@ describe("typespec-client-generator-core: multipart types", () => {
...
+        op basic1(@header contentType: "multipart/form-data", @body body: MultiPartRequest): NoContentResponse;
+        @post
+        @route("/basic2") 
+        op basic2(@header contentType: "multipart/form-data", @body body: MultiPartRequest): NoContentResponse;
+      `
+    );
+    deepEqual(runner.context.diagnostics.length, 0);
+    const address = runner.context.sdkPackage.models.find((x) => x.name === "Address");
+    ok(address);
+    deepEqual(address.usage, UsageFlags.Input);
```
Comment (msyyc) [inline packages/typespec-client-generator-core/test/types/multipart-types.test.ts:86]: Since `usage.Json` is calculated recursively from regular body. If it is not used in regular body, the model usage will not be assigned to `Json`.

File: packages/typespec-client-generator-core/test/types/multipart-types.test.ts (hunk 3 above)
Comment (msyyc) [reply packages/typespec-client-generator-core/test/types/multipart-types.test.ts:100]: No. You can see https://github.com/Azure/typespec-azure/pull/1511/files#r1758270473

File: packages/typespec-client-generator-core/src/types.ts (hunk 1 above)
Comment (iscai-msft) [reply packages/typespec-client-generator-core/src/types.ts:1668]: can you clean up this logic so it's easier to read? i.e. can you make vars like `isXml = sdkType.usage & UsageFlags.Xml > 0` because right now it's really hard to follow what the logic is trying to do here

Comment (tadelesh) [reply packages/typespec-client-generator-core/src/types.ts]: @iscai-msft i've refined the logic, could you review again and help to merge?

Comment (iscai-msft) [reply packages/typespec-client-generator-core/src/types.ts]: do we want this to merge into the release branch or into the main branch?

**Review by iscai-msft — APPROVED** [review]

File: packages/typespec-client-generator-core/src/types.ts (hunk 1 above)
Comment (tadelesh) [reply packages/typespec-client-generator-core/src/types.ts]: main is ok.
